src/widgetastic/widget.py

```diff
@@ -1514,14 +1537,42 @@ def fill(self, value):
         if isinstance(value, dict):
             if self.assoc_column_position is None:
                 raise TypeError('In order to support dict you need to specify assoc_column')
-            return any(
-                self.row((self.assoc_column_position, key)).fill(fill_value)
-                for key, fill_value
-                in six.iteritems(value))
+            changed = False
+            for key, value in six.iteritems(value):
+                try:
+                    row = self.row((self.assoc_column_position, key))
+                except IndexError:
+                    row_id = self.add_row_handler()
+                    if row_id is None:
+                        raise TypeError('Could not add a row into table {!r}'.format(self))
+                    row = self[row]
+                if row.fill(value):
+                    changed = True
+            return changed
         else:
             if not isinstance(value, (list, tuple)):
                 value = [value]
-            return any(row.fill(value) for row, value in zip(self, value))
+            rows = list(self)
+            value_row_mapping = list(zip(value, rows))
+            extra_row_values = value[len(value_row_mapping):]
+
+            changed = any(row.fill(value) for value, row in value_row_mapping)
+            for extra_value in extra_row_values:
+                row_id = self.add_row_handler()
+                if row_id is None:
+                    raise TypeError('Could not add a row into table {!r}'.format(self))
+                row = self[row_id]
+                if row.fill(extra_value):
+                    changed = True
+            return changed
+
+    @property
+    def row_count(self):
+        return len(self)
+
+    def add_row_handler(self):
```

Comment:
This will handle adding new rows. Will be overridden by the specific implementation.

I haven't figured out yet the best way to handle deleting rows as a part of filling